lib/writeGltf.js

```diff
@@ -0,0 +1,85 @@
+'use strict';
+var fs = require('fs');
+var path = require('path');
+var Cesium = require('cesium');
+var defined = Cesium.defined;
+var dataUri = require('datauri');
+var async = require('async');
+var mkdirp = require('mkdirp');
+
+module.exports = writeGltf;
+
+function writeGltf(gltf, outputPath, isEmbedded, createDirectory, callback) {
+    //Create the output directory if specified
+    if (createDirectory) {
+        outputPath = path.join(path.dirname(outputPath), 'output', path.basename(outputPath));
+        mkdirp.sync(path.dirname(outputPath));
+    }
+    
+    var shaders = gltf.shaders;
+    //Write out the source objects, followed by the entire gltf object
+    if (defined(shaders)) {
+        var shaderIds = Object.keys(shaders);
+        async.each(shaderIds, function(shaderId, asyncCallback) {
+            var shader = shaders[shaderId];
+            if (defined(shader.extras) && defined(shader.extras.source)) {
+                var source = shader.extras.source;
+                delete shader.extras.source;
+
+                var filePath = path.dirname(outputPath);
+                writeSource(filePath, shaderId, shader, source, isEmbedded, asyncCallback);
+            }
+        }, function(err) {
+            if (err) {
+                if (callback) {
+                        process.nextTick(function() {

```

Comment:
I don't think I was clear about this before, if we call a callback from within a callback (like this case here since we are calling `callback` inside the callback passed to `async.each` then we do not need to use `proces.nextTick` since the first callback should have handled it.  I _think_ this is the best practice to keep the code from being too verbose.